parser/src/moves.rs

```diff
@@ -0,0 +1,173 @@
+use lexer::token::TokenType::{EndOfFile, Space};
+use lexer::token::{Token, TokenType};
+
+///defines a way to move along a ParserCursor.
+pub trait Move {
+    /// Returns
+    /// * `Some<usize>` - if the move succeeded, where the wrapped `usize` is the position where this move ended.
+    /// * `None` - if the move did not succeed (prerequisites not satisfied)
+    /// # Arguments
+    /// `None` if the move did not take effect.
+    ///* `at` - get token at given position
+    ///* `pos` - the position in ParserCursor at beginning of the move
+    fn apply<'a, F>(&self, at: F, pos: usize) -> Option<usize>
+        where
+            F: Fn(usize) -> Token<'a>;
+}
+
+///Defines operations over a Move struct.
+pub(crate) trait MoveOperations<'a, This: Move> {
+    ///Used to chain `This` move with `other` move.
+    /// returns a move that will first execute this move then other one only if this first succeeded.
+    fn and_then<B: Move>(self, other: B) -> AndThenMove<This, B>;
+
+    ///Used to bind `This` move with `other` move.
+    /// returns a move that will first execute this move then the other one.
+    fn then<B: Move>(self, other: B) -> ThenMove<This, B>;
+}
+
+impl<'a, A: Move> MoveOperations<'a, A> for A {
```

Comment:
```suggestion
impl<A: Move> MoveOperations<A> for A {
```
Unnecessary qualified lifetimes